File: convert_images.js
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';

const dir = 'public';

function walk(directory) {
  let results = [];
  const list = fs.readdirSync(directory);
  list.forEach(file => {
    file = path.join(directory, file);
    const stat = fs.statSync(file);
    if (stat && stat.isDirectory()) {
      results = results.concat(walk(file));
    } else {
      results.push(file);
    }
  });
  return results;
}

const files = walk(dir);

files.forEach(file => {
  if (file.match(/\.(png|jpg|jpeg)$/i)) {
    const webpPath = file.replace(/\.(png|jpg|jpeg)$/i, '.webp');
    sharp(file)
      .webp({ quality: 80 })
      .toFile(webpPath)
      .then(() => console.log(`Converted ${file} to ${webpPath}`))
      .catch(err => console.error(`Error converting ${file}: ${err}`));
  }
});

File: convert_images.py
import os
from PIL import Image

def convert_to_webp(directory):
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.lower().endswith(('.png', '.jpg', '.jpeg')):
                filepath = os.path.join(root, file)
                try:
                    img = Image.open(filepath)
                    webp_path = os.path.splitext(filepath)[0] + '.webp'
                    img.save(webp_path, 'WEBP', quality=85)
                    print(f"Converted {filepath} to {webp_path}")
                except Exception as e:
                    print(f"Failed to convert {filepath}: {e}")

if __name__ == "__main__":
    convert_to_webp(r"e:\PROJECTS\threejs_portfolio\public")
